### ISSUE TITLE
Hidden top bar opens but then immediately becomes invisible if active window is currently maximized (top bar is still there; pressing buttons on it works)

### ISSUE THREAD

**Issue by mlncn** [mlutfy/hidetopbar#372]
The problem is perhaps the top bar being hidden *too* well, heh.  But really everything works perfectly, with animations on one can see the top bar showing up as expected, but then once the animation is done it disappears instantly.  (With animations off, set to zero, the same thing is happening, an invisible top bar becomes present and buttons on it can be pressed, but there is no visual cue it ever showed up.)

Thought this was specific to Firefox because that's generally the only maximized application, but can reproduce even with system settings dialogs if i maximize them.

It is a problem of course not to be able to see the top bar to know what time it is or what buttons are available where, and it is also hard to know when the bar is still there, so it being there invisible interferes with pressing top tabs on the open application.

I think the development of this problem coincided with the second-most-recent update of Gnome.

Currently on Gnome 42.2 and the Wayland windowing system.  Debian GNU/Linux bookworm/sid.

**Comment by GrigDavid** [comment]
Same here. It also caused dash to dock for cosmic with autohide enabled have the same problem. I'll disable this extension until I find a fix

**Comment by mumblingdrunkard** [comment]
Experiencing the same issue. Both Dash to Dock for COSMIC and Hide Top Bar started acting this way recently.
It must've been a GNOME 42 update that broke it, but I can't pinpoint exactly which one.

**Comment by mumblingdrunkard** [comment]
Quick update; A little testing shows that Dash to Dock for COSMIC stops acting up if I disable Hide Top Bar.
The same is not true the other way around.

**Comment by osieltorres** [comment]
Same happening to me, using Pop!_OS 22.04, both Dock and Top-bar behaves that way. The first started blinking after cursor passes through icons, and the second hides after approximately one second.

**Comment by GrigDavid** [comment]
this happens in some apps that have not regular title bar, like firefox or gnome software. If dash-to-dock flickers, disabling hide-top-bar should fix dash-to-dock

**Comment by mayankwadhwani** [comment]
I think this should be fixable if we put an always on top functionality on the panel. Is there any code pointer for the same on which we can try if anyone is aware ?

**Comment by FractalMachinist** [comment]
Re @GrigDavid I also see this flickering only when the fullscreen app doesn't use system titles & borders. Ubuntu 22.04 + GNOME 42.2 (Wayland) + Chrome 104.0.5112.79 + Hide Top Bar 108 + Ubuntu Dock 71 (no dash-to-dock)

**Comment by AAcatLeader** [comment]
I'm having this same issue on Manjaro Gnome 42.4 Wayland, When this extension (vr. 108) is enabled. Any full sized applications will cause both the top bar and the dock to appear momentarily. followed by flickering before becoming invisible (but still clickable).

**Comment by sdwolfz** [comment]
Same problem, Manjaro Gnome. I'm building the extension package using this PKGBUILD here: https://gitlab.com/sdwolfz/pkgbuilds/-/blob/master/gnome-shell-extension-autohidetopbar/PKGBUILD

Using latest commit 74cc28b5f2be11a2e3157689610ef7dd19b6015d (which was supposed to fix this).

I'm thinking: does the plugin store some form of state somewhere? Maybe that needs to be cleared since the above commit... since it does not happen to everybody maybe uninstalling, clearing whatever, then installing again would work :shrug:

**Comment by YueRen** [comment]
I think this is a wider issue and not an issue of the extension itself. I have the exact same issue with the "Auto-hide the Dock" option in the appearance setting menu of Ubuntu 22.04 (toggle option + maximize Firefox + move mouse to side of screen = dock appears and disappears).

**Comment by sdwolfz** [comment]
@tuxor1337 can you please post a list of installed extensions you have on your system? Maybe it's an interaction with another that we might have and you don't. (I'll post my list later today when I arrive home)

**Comment by YueRen** [comment]
In Ubuntu 22.04, the extension also messes up the "Auto-hide the Dock" option in Ubuntu's appearance setting (Firefox maximized + mouse on edge of screen = dock appears and disappears).

**Comment by tuxor1337** [comment]
> @tuxor1337 can you please post a list of installed extensions you have on your system? Maybe it's an interaction with another that we might have and you don't. (I'll post my list later today when I arrive home)

I don't use and never used any other extensions. Additionally, note that I'm on Fedora where you start with a vanilla GNOME that doesn't come with default extensions like a dock etc.

**Comment by sdwolfz** [comment]
Here's what I have:
* `ls ~/.local/share/gnome-shell/extensions/` - none
* `gsettings get org.gnome.shell enabled-extensions`:
    ```
    ['pamac-updates@manjaro.org', 'dash-to-dock@micxgx.gmail.com', 'gnome-ui-tune@itstime.tech', 'x11gestures@joseexposito.github.io', 'ding@rastersoft.com', 'appindicatorsupport@rgcjonas.gmail.com', 'hidetopbar@mathieu.bidon.ca']
    ```
* `ls /usr/share/gnome-shell/extensions/`:
    ```
    appindicatorsupport@rgcjonas.gmail.com                       gnome-ui-tune@itstime.tech                                         places-menu@gnome-shell-extensions.gcampax.github.com
    apps-menu@gnome-shell-extensions.gcampax.github.com          gsconnect@andyholmes.github.io                                     screenshot-window-sizer@gnome-shell-extensions.gcampax.github.com
    arcmenu@arcmenu.com                                          hidetopbar@mathieu.bidon.ca                                        user-theme@gnome-shell-extensions.gcampax.github.com
    auto-move-windows@gnome-shell-extensions.gcampax.github.com  launch-new-instance@gnome-shell-extensions.gcampax.github.com      window-list@gnome-shell-extensions.gcampax.github.com
    dash-to-dock@micxgx.gmail.com                                material-shell@papyelgringo                                        windowsNavigator@gnome-shell-extensions.gcampax.github.com
    dash-to-panel@jderose9.github.com                            native-window-placement@gnome-shell-extensions.gcampax.github.com  workspace-indicator@gnome-shell-extensions.gcampax.github.com
    ding@rastersoft.com                                          nightthemeswitcher@romainvigier.fr
    drive-menu@gnome-shell-extensions.gcampax.github.com         pamac-updates@manjaro.org
```

Part of Manjaro's Defaults... I'll disable each extra until It works I guess...

**Comment by AAcatLeader** [comment]
Dash to dock recently updated, I tested out this extension to see and it seems to work perfectly now.

**Comment by AAcatLeader** [comment]
> 

never mind, stopped working as before

**Comment by suppai** [comment]
I can confirm this behavior on a brand new install of Fedora 36 with no other extensions if it helps @tuxor1337 --- installing Dash to Dock makes the problem worse (and breaks Dash to Dock as well), but even without it the problem still occurs.

**Comment by bruncarv1** [comment]
It hapens using wps text editor ou firefox maximized, wiht other app goes well

**Comment by Nugine** [comment]
https://github.com/flameshot-org/flameshot/issues/2590#issuecomment-1177189932
It also solves this issue.

**Comment by sdwolfz** [comment]
Interesting, so It's a Wayland issue, makes sense since I use Wayland. 

@tuxor1337 are you on Xorg? and can you debug this on Wayland?

Side note: I don't know how to debug this myself, any hints?

**Comment by tuxor1337** [comment]
@sdwolfz As mentioned above, I'm on Wayland and can't reproduce this.

**Comment by bruncarv1** [comment]
Yes, for me switching wayland to xorg solves the problem! But I think it shoud be fixed, wont you all agree?
tks

**Comment by sdwolfz** [comment]
I tried debugging this with some extra debug statements, see here: https://gitlab.com/sdwolfz/hidetopbar/-/compare/master...flickering-fix

The only difference I see in the arguments to "_handleMenus" is an extra...
```
              {
                "id": 39,
                "name": "notify::sensitive",
                "disconnected": false
              },
              {
                "id": 40,
                "name": "notify::sensitive",
                "disconnected": false
              }
```
...in the `0 >   statusArea >   appMenu >   menu >   _windowSection >   _signalConnections` section for when it's working properly (that part is missing when using firefox fullscreen, but present when using gnome terminal fullscreen for example).

Here are the full log files:

[log-broken.log](https://github.com/tuxor1337/hidetopbar/files/9739025/log-broken.log)
[log-working.log](https://github.com/tuxor1337/hidetopbar/files/9739026/log-working.log)

That's as much as I'm able to investigate right now.

**Comment by diogodomanski** [comment]
> Interesting, so It's a Wayland issue, makes sense since I use Wayland.
> 
> @tuxor1337 are you on Xorg? and can you debug this on Wayland?
> 
> Side note: I don't know how to debug this myself, any hints?

I've experienced this issue only with Wayland. Xorg all works fine

**Comment by sdwolfz** [comment]
@tuxor1337 you can't reproduce this with Firefox on Ubuntu because over there, the Firefox package is installed via Snap. Similarly I can't reproduce this with the Snap package on Manjaro Wayland.

But if you would install Firefox fro ma native package instead you would get the same behavior.

Here is an explanation of the behaviour:
- You start up Firefox, then maximize it, top bar hides
- You scroll to the top, the top bar animation plays drawing on top of Firefox
- The animation ends, Firefox draws itself on top of the top bar, so it's now invisible, but the top bar elements are still logically above the browser window.
- Move the mouse inside the top bar, it becomes visible again while the motion persists.
- Stop moving the mouse, Firefox again draws itself on top of the top bar.
- Move the mouse outside the top bar, it becomes visible again, and while the animation lasts it's drawn on top.
- Finally the top bar is hidden for real.

What I believe happens, is that there is something like a `z-index` issue for the two. Both Firefox and the top bar are at the same level and whichever one of them happens to draw itself on top wins, that's why you get the start/end animations and visibility while moving the mouse.

**Comment by tuxor1337** [comment]
@sdwolfz Thanks for your attempt of an explanation! As mentioned above, I'm on Fedora with Wayland and can't reproduce this. snap is not installed on this machine.

**Comment by sdwolfz** [comment]
OK, my quest continues

**Comment by sdwolfz** [comment]
With the changes I made in https://github.com/tuxor1337/hidetopbar/pull/390 I've debugged this further:

[notworks.txt](https://github.com/tuxor1337/hidetopbar/files/9842740/notworks.txt)
[works.txt](https://github.com/tuxor1337/hidetopbar/files/9842742/works.txt)

As you can see in the logs there, when it works well, with gnome terminal fullscreen, the logs always start with:
```
PanelVisibilityManager show(mouse-enter)
```

But when firefox fullscreen is in play, it starts with:
```
Intellihide _checkOverlapSignal function(): name = "notify::focus-app" | args = [{},{}]
```

**Comment by Wemmy0** [comment]
Yeah this is happening with other apps to on my laptop.
The app Obsidian MD also has this issue. It's only happening in wayland and it also effects dash to dock extension which experiences the same issue when Hide Top Bar is enabled.

**Comment by yo-ranzz** [comment]
Hey, here is a video that I recorded showing the issue. Might help. I'm on fedora 37 beta but it still occurs in F36. Also, for some reason it doesn't show in any screen recording software so I used my phone. 
https://github.com/yo-ranzz/Hide-top-bar-issue/blob/main/README.md

**Comment by Wemmy0** [comment]
I tried this on a brand-new fedora install, and it happened immediately after installing the extension and restarting my laptop.

**Comment by CO-1** [comment]
TLDR: is a gnome-shell/mutter bug, my extension: https://github.com/CO-1/gnome-shell-extension-disableunredirect  should workaround it but may have performance penalty for gaming (and may want to disable it while gaming). 

This is a long standing issue. It happened with vscode and the workaround was to disable custom title bar, but now it happens with firefox and even gnome-extensions-app.

For a while now I used a workaround discovered accidentally: make a window "Always on Top" or "Always on Visible Workspace" but that would not always work (turns out animations had to be disabled in gnome)
After gnome shell 43 this workaround did not work at all. I thought it was a good starting point to investigate the root cause and maybe fix this for good.
After bisecting turned out that commit ee09c5c85312f571e14abe69bb6674a361c16d65 would prevent workaround to work.
Further testing reveled that Clutter.Actor.prototype.set_easing_duration override in js/ui/environment.js was the culprit and changing it to call origSetEasingDuration with a min value of 1 would allow workaround to still function.

At this point I was pretty sure it is a gnome-shell or mutter bug.
To prove that I made a simple extension that just removes top bar (with none of the features of hidetopbar).
Tested with dashtodock autohide set to "Always on top" and as soon as topbar is gone bug is there, dashtodock would disapear and hovering over its position would reveal it briefly, clicking would open the application.
The behavior is still there after disabling test extension restoring topbar.

While trying to understand gnome-shell easing code I noticed disable_unredirect_for_display / enable_unredirect_for_display are called  before/after animations. I gave it a try and after disable_unredirect_for_display everything works so I made a extension for that:
https://github.com/CO-1/gnome-shell-extension-disableunredirect

I may be wrong but my understanding is that unredirect allows a window to skip compositor on full screen, disabling unredirect always forces compositor for all windows. This may have performance impact on games so one may want to stop disableunredirect while gaming.

**Comment by sdwolfz** [comment]
@tuxor1337 @CO-1 Maybe it's a silly thing to say and I'm not understanding it correctly, but how about we:

1. `disable_unredirect_for_display` if intellihide is active and topbar transitions from hidden to visible.
2. keep `disable_unredirect_for_display` while topbar is visible
3. `enable_unredirect_for_display` once the topbar transitions from visible to hidden and it was previously in state 1.

I'm probably missing some edge cases related to fullscreen/non fullscreen apps. But could this work? I don't think it's going to cause any issues for games since you go back to the optimised state once the bar is hidden again.

**Comment by CO-1** [comment]
@sdwolfz I suspect unredirect (compositor bypass ) is being performed for maximized windows even though I've only read about full screen.

You are right,  a better solution would be calling disable_unredirect_for_display on topbar show and enable_unredirect_for_display on topbar hide when fullscreen or maximized window. This would have to be implemented in hidetopbar and dashtopanel and any other extension that removes topbar.
Best solution would be for someone to fix this bug in gnome.

It was easier for me to just make a extension, it solves the problem for all other extensions that remove top bar. Besides composition performance penalty should only have a impact on gaming.

**Comment by eyrent** [comment]
@CO-1 I tried the fix extension you made and I works well!!! Thank you so much! Do you have any ideas about [this](https://github.com/tuxor1337/hidetopbar/issues/396). Unfortunately that issue is still there after the patch. Cheers!

**Comment by CO-1** [comment]
@eyrent #396 annoys me too but did not have time and patience to debug it.
It only affects 1 window, the focused one, so I use Win+DownArrow, Win+UpArrow to unmaximize/maximize the window after resume.
